### PR TITLE
AC-27: forget password - SendEmail function

### PR DESCRIPTION
**Pull request by miaozhaod**
### Provide description as to what you have done in this pr.
- Create the send email function for forget password

### Provide screenshots or videos that the function is working.
Screen Shot 2022-05-29 at 8 26 54 am

### Provide screenshots of no warning, error in console.
See above

## That the feature you have worked on has its...
- [ ] relevant information updated in README.MD
- [ ]  swagger doc created/updated so that other developers can read the related documentation.
- [ ]  logging included/updated so that we can debug errors when the app is live in the UAT environment.
- [ ]  unit tests created/updated so that we can integrate this new release and sleep without worries at night/day.

When you finished, you will want to make sure you have...
- [X]  No merge conflict with develop branch.
- [X]  Check PR title is correct, (i.e [AC-xx: Feature] Create pull request template).
- [X]  Check your app will run without crashes.
- [ ]  Check if there's hard coded values.
- [X]  No //eslint-disable-next-line no-unused-vars, unless otherwise specified.
- [X]  No unnecessary spaces between lines of code.
- [X]  Check you are using camelCase && PascalCase correctly.
- [X]  Check you named your code properly.
- [X]  Remove unused comments.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -35,8 +35,11 @@
     "bcrypt": "^5.0.1",
     "class-transformer": "^0.5.1",
     "class-validator": "^0.13.2",
+    "date-fns": "^2.28.0",
     "graphql": "^16.4.0",
+    "graphql-scalars": "^1.17.0",
```
Comment (NorrisWu0) [inline package.json]: Don't think we use this anymore?

File: src/main.ts [hunk 2]
```diff
@@ -1,11 +1,14 @@
 import { NestFactory } from '@nestjs/core';
 import { Logger, ValidationPipe } from '@nestjs/common';
 import { ConfigService } from '@nestjs/config';
-
+import { PrismaClient } from '@prisma/client';
```
Comment (NorrisWu0) [inline src/main.ts]: This is probably an unused line

File: src/resource/auth/dto/forget-password.input.ts [hunk 3]
```diff
@@ -0,0 +1,9 @@
+import { InputType, Field } from '@nestjs/graphql';
+import { IsNotEmpty } from 'class-validator';
+
+@InputType()
+export class ForgetPasswordInput {
+  @Field()
+  @IsNotEmpty()
+  username: string;
```
Comment (NorrisWu0) [inline src/resource/auth/dto/forget-password.input.ts]: Maybe we should use `email` instead

File: src/resource/user/user.service.ts [hunk 4]
```diff
@@ -17,4 +21,23 @@ export class UserService {
   async findByUsername(_username: string) {
     return this.prisma.user.findFirst({ where: { username: _username } });
   }
+
+  async createPasswordResetLink(_id: string) {
+    const { createHmac, randomBytes } = await import('crypto');
+    const secret = process.env.CRYPTO_SECRET;
+    const token = randomBytes(32).toString('hex');
+    const resetToken = createHmac('sha256', secret).update(token).digest('hex');
```
Comment (NorrisWu0) [inline src/resource/user/user.service.ts]: Since we are not really reusing the secret here, probably can shorten this to use `process.env.CTOPTO_SECRET` directly. Unless there's a different reason to keep it. 👌 

Or it can be there to make the code readable? which makes sense too.

File: src/resource/user/user.service.ts [hunk 5]
```diff
@@ -17,4 +21,23 @@ export class UserService {
...
+    const resetTokenExpires = addHours(new Date(), 1).toISOString();
+    console.log(resetTokenExpires);
```
Comment (NorrisWu0) [inline src/resource/user/user.service.ts]: So instead of `console.log`, let's use 
```
const { Logger } from '@nestjs/common'

Logger.log()
```

No particular reason, it just looks nicer and has more info.

Comment (NorrisWu0) [reply src/resource/user/user.service.ts]: Also, we don't want to log the expiry out. We just need to log the operation has success or log error if there's anything wrong with it.

File: src/utils/sendEmail.ts [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+
+  await transporter.sendMail({
+    from: '"🍍 A-COMOSUS 🍍" <project.a.comosus@gmail.com>',
+    to: email,
+    subject: `Reset Password`,
+    text: emailContent,
+    html: emailContent,
+  });
+
+  console.log('Email sent successfully!');
```
Comment (NorrisWu0) [inline src/utils/sendEmail.ts]: Same goes with my last comment.

File: package.json (hunk 1 above)
Comment (miaozhaod) [reply package.json]: this has been removed

File: src/main.ts (hunk 2 above)
Comment (miaozhaod) [reply src/main.ts]: this has been removed

File: src/resource/user/user.service.ts (hunk 4 above)
Comment (miaozhaod) [reply src/resource/user/user.service.ts]: This has been updated

File: src/resource/user/user.service.ts (hunk 5 above)
Comment (miaozhaod) [reply src/resource/user/user.service.ts]: This has been updated

File: src/utils/sendEmail.ts (hunk 6 above)
Comment (miaozhaod) [reply src/utils/sendEmail.ts]: This has been updated

File: src/resource/auth/dto/forget-password.input.ts (hunk 3 above)
Comment (miaozhaod) [reply src/resource/auth/dto/forget-password.input.ts]: this has been updated

**Review by NorrisWu0 — APPROVED** [review]
